### PR TITLE
Add PeerConnection.GracefulClose

### PR DESCRIPTION
**Pull request by edaniels**
Take 2: This is a more thorough attempt at closing gracefully. If anything, every path that graceful touches becomes opt-in, so it shouldn't affect any existing users and it makes it safe for backporting.

GracefulClose/Stop added to:
- PeerConnection
- DataChannel
- ICEGatherer
- ICETransport
- operations

### CODE REVIEW

**Comment by edaniels** [comment]
First test run had https://github.com/pion/webrtc/issues/2848 but verified it happens on `master`

File: peerconnection.go [hunk 1]
```diff
@@ -2092,13 +2096,41 @@ func (pc *PeerConnection) writeRTCP(pkts []rtcp.Packet, _ interceptor.Attributes
 	return pc.dtlsTransport.WriteRTCP(pkts)
 }
 
-// Close ends the PeerConnection
+type shouldGracefullyClose bool
+
+const (
+	shouldGracefullyCloseNo  shouldGracefullyClose = false
+	shouldGracefullyCloseYes shouldGracefullyClose = false
```
Comment (Sean-Der) [inline peerconnection.go:2103]: should this be true?

Comment (Sean-Der) [reply peerconnection.go:2103]: It's a matter of style, but this pattern is easier/more popular IMO

```
return pc.close(false /* shouldGracefullyClose */)
```

File: datachannel.go [hunk 2]
```diff
@@ -40,6 +40,8 @@ type DataChannel struct {
 	readyState                 atomic.Value // DataChannelState
 	bufferedAmountLowThreshold uint64
 	detachCalled               bool
+	readLoopActive             chan struct{}
+	isUserClosed               bool
```
Comment (Sean-Der) [inline datachannel.go:44]: `isGracefulClose`? I see this set via `Close` or `GracefulClose` only

File: peerconnection.go (hunk 1 above)
Comment (edaniels) [reply peerconnection.go:2103]: yep and sure good idea

File: datachannel.go (hunk 2 above)
Comment (edaniels) [reply datachannel.go:44]: good idea, more specific :)

File: internal/mux/mux.go [hunk 3]
```diff
@@ -129,6 +129,10 @@ func (m *Mux) readLoop() {
 		}
 
 		if err = m.dispatch(buf[:n]); err != nil {
+			if errors.Is(err, io.ErrClosedPipe) {
```
Comment (edaniels) [inline internal/mux/mux.go:132]: I don't see a reason to have this spuriously log when we are the ones closing the packetio.Buffer.

**Review by Sean-Der — APPROVED** [review]

**Pull request merged**